PMSCore/User.cs

```diff
@@ -0,0 +1,247 @@
+namespace PMSCore;
+
+public abstract class User : IUser
+{
+    protected User(string userId, HashSet<Permission> userPermissions)
+    {
+        UserId = userId;
+        UserPermissions = userPermissions;
+        UserReport = string.Empty;
+    }
+
+    private void AppendToReport(string message)
+    {
+        Console.WriteLine(message);
+        UserReport += $"[{DateTime.UtcNow}] [{UserId}] [{message}]\n";
+    }
+
+    public UserOperationResult CreateEntity(EntityType type, string entityId, string? entityDescription = null)
+    {
+        switch (type)
+        {
+            case EntityType.Project:
+                if (!UserPermissions.Contains(Permission.CanCreateProject))
+                {
+                    AppendToReport("Not enough permissions to create a project");
+                    return UserOperationResult.Failed;
+                }
+
+                break;
+            case EntityType.Task:
+                if (!UserPermissions.Contains(Permission.CanCreateTask))
+                {
+                    AppendToReport("Not enough permissions to create a task");
+                    return UserOperationResult.Failed;
+                }
+
```

Comment:
if (!UserPermissions.Contains(Permission.CanCreateTask)) - maybe it would be better as:

if (!HasPermission(Permission.CanCreateTask)) - to encapsulate into a helper method to avoid redundancy because UserPermissions.Contains(permission) is used more than once in User.cs

Reply:
Reused CheckUserPermission already implemented in User

Reply:
That's a good point, CheckUserPermission already provides encapsulation.